std/container/fixed_array.d

```diff
@@ -0,0 +1,504 @@
+module std.container.fixed_array;
+
+import core.exception, std.algorithm, std.conv, std.exception, std.range,
+        std.traits, std.typecons;
+
+public import std.container.util;
+version(unittest) import std.stdio, std.container;
+
+/**
+Array type that uses a $(XREF2 range, isRandomAccessRange, random access range)
+as fixed storage location. No memory is ever allocated. Throws if not enough
+space is left for an operation to succeed.
+*/
+struct FixedArray(Store)
+    if(isRandomAccessRange!Store && hasLength!Store)
+{
+
+/**
+    $(XREF2 range, isRandomAccessRange, random access range) that provides
+    the same range primitives as $(D Store).
+
+    $(LREF opIndex) and $(LREF opSlice) return values of this type.
+    */
+    alias Range = Store;
+
+/**
+ * The type of the elements that can be stored in this
+ * FixedArray.
+ */
+    alias Element = ElementType!Range;
+
+    private {
+
+        static struct Data {
+            Store store;
+            size_t length;
+        }
+
+        alias Payload = RefCounted!(Data, RefCountedAutoInitialize.yes);
+        Payload payload;
+
+        @property
+        ref inout(Store) store() inout { return payload.store; }
+
+    }
+
+
+/**
+    Constructor taking a random access range of type $(D Store).
+
+    Params:
+        store = $(XREF2 range, isRandomAccessRange, random access range) used as a store
+        initialLength = the initial length of the FixedArray. If set two $(D 0),
+                        all elements in store will be overwritten and if set to
+                        $(D range.length) no elements can be inserted into the
+                        $(D FixedArray) until some are removed.
+     */
+    this(Store store, size_t initialLength = size_t.max)
+    {
+
+        payload.store = move(store);
+        length = min(initialLength, store.length);
+    }
+
+
+/**
+Comparison for equality.
+
+    Two FixedArrays are equal, if they have the same length and
+    corresponding elements are equal.
+     */
+    bool opEquals(FixedArray rhs)
+    {
+        return opEquals(rhs);
+    }
+
+    /// ditto
+    bool opEquals(ref FixedArray rhs)
+    {
+        if(length != rhs.length)
+            return false;
+        return equal(store[0 .. length], rhs.store[0 .. length]);
+    }
+
+static if(is(typeof(store.dup)))
+{
+/**
+Duplicates the container. The elements themselves are not transitively
+duplicated. The underlying store must support $(D .dup) as well.
+
+Complexity: $(BIGOH n).
+     */
+    @property FixedArray dup()
+    {
+        return FixedArray(store.dup, length);
+    }
+}
+
+
+
+/**
+Property returning $(D true) if and only if the container has no
+elements.
+
+Complexity: $(BIGOH 1)
+     */
+    @property bool empty() const
+    {
+        return length == 0;
```

Comment:
It seems to me an uninitialized array (one with a null RefCounted) should just be empty without triggering an allocation of the RefCounted: `return payload is payload.init || length == 0;`